plugins/scroll-options/src/ScrollBlockDragger.js

```diff
@@ -121,5 +128,96 @@ export class ScrollBlockDragger extends Blockly.BlockDragger {
     this.updateToolboxStyle_(true);
 
     this.dragDelta_ = currentDragDeltaXY;
+
+    // Added
+    this.stopAutoScrolling();
+  }
+
+  /**
+   * May scroll the workspace as a block is dragged.
+   * If a block is dragged near the edge of the workspace, this method will
+   * cause the workspace to scroll in the direction the block is being dragged.
+   * The workspace will not resize as the block is dragged. The workspace should
+   * appear to move out from under the block, i.e., the block should stay under
+   * the user's mouse.
+   * @param {!Blockly.utils.Coordinate} newLoc New coordinate the block is being
+   *     dragged to.
+   */
+  scrollWorkspaceWhileDragging(newLoc) {
+    const SCROLL_DIRECTION_VECTORS = {
+      top: new Blockly.utils.Coordinate(0, 1),
+      bottom: new Blockly.utils.Coordinate(0, -1),
+      left: new Blockly.utils.Coordinate(1, 0),
+      right: new Blockly.utils.Coordinate(-1, 0),
+    };
+    // TODO(maribethb): I just made this up, pick a better one
+    // and make this configurable.
+    const SCROLL_SPEED = 0.4;
+
+    const candidateScrolls = [];
+    let overallScrollVector = new Blockly.utils.Coordinate(0, 0);
+
+    // Get ViewMetrics in workspace coordinates.
+    const metrics = this.workspace_.getMetricsManager().getViewMetrics(true);
+
+    // TODO(maribethb): Add fancier logic based on how far out of bounds the
+    // block is held.
+
+    // See Blockly.MetricsManager for more information on the metrics used.
+    // In particular, it uses workspace coordinates where the top and left
+    // of the workspace are negative.
+    // More than one scroll vector may apply, for example if the block is
+    // dragged to a corner.
+    if (newLoc.y < metrics.top) {
+      const scrollVector = SCROLL_DIRECTION_VECTORS['top'].scale(SCROLL_SPEED);
+      candidateScrolls.push(scrollVector);
+    }
+    if (newLoc.y > metrics.top + metrics.height) {
+      const scrollVector =
+          SCROLL_DIRECTION_VECTORS['bottom'].scale(SCROLL_SPEED);
+      candidateScrolls.push(scrollVector);
+    }
+    if (newLoc.x < metrics.left) {
+      const scrollVector = SCROLL_DIRECTION_VECTORS['left'].scale(SCROLL_SPEED);
+      candidateScrolls.push(scrollVector);
+    }
+    if (newLoc.x > metrics.left + metrics.width) {
+      const scrollVector =
+          SCROLL_DIRECTION_VECTORS['right'].scale(SCROLL_SPEED);
+      candidateScrolls.push(scrollVector);
+    }
+
+    // Get the overall scroll direction vector (could scroll diagonally).
+    // CDO reduces down to just one vector per direction from all the possible
+    // ones they generate. Currently we just have one per direction so we don't
+    // need to do anything else.
```

Comment:
This was more of a todo/note to myself :) I don't think it'll be in here after the extra dragging logic is added.

They have more than one per direction because they have several conditions that could be fulfilled simultaneously. For example, they could have "scroll fast if the block is 30 px past the edge of the ws" *and* "scroll slow if the mouse is within 10px of the edge". Both of those conditions could be met simultaneously, and you don't want to just add them together to scroll even faster. You just want the fastest one to apply. So you'd have to select the fastest vector in each direction, and then add the resulting one vector from each direction together. Here, I'm only adding the directions together, hence the reminder to condense down to one per direction before adding.